### ISSUE TITLE
Debian 9 / Mariadb 10.1: install fails when creating session tables - "1071 Specified key was too long"

### ISSUE THREAD

**Issue by rotanid** [passbolt/passbolt_api#163]
i tried to set up passbolt 1.6.3 on a debian server running PHP 7.0.19-1 and MariaDB 10.1.26-0+deb9u1 .
but when running 
`app/Console/cake install --no-admin`
as the user, the routine seems to run fine but actually fails in the middle with
```
The following table(s) will be created.
cake_sessions
Creating table(s).
cake_sessions: SQLSTATE[42000]: Syntax error or access violation: 1071 Specified key was too long; max key length is 767 bytes
End create.
passbolt session table deployed
```
did i miss something?

**Comment by psy-q** [comment]
I have the same issue, and after installing healthcheck will fail with "No table found". It installs the table but the SQLSTATE error mention occurs again. I can see the tables and there is also content (e.g. in the `roles` table), but Passbolt insists that "No table found".

**Comment by cedricalfonsi** [comment]
I just tried with different versions of mardiadb (10.1.26 / 10.2.8) and I'm not able to reproduce your issue.

Googling the error message I found people with the same kind of problem, and it could be related to the database engine.

> Prefix support and lengths of prefixes (where supported) are storage engine dependent. For example, a prefix can be up to 767 bytes long for InnoDB tables or 3072 bytes if the innodb_large_prefix option is enabled. For MyISAM tables, the prefix limit is 1000 bytes. The NDB storage engine does not support prefixes (see Section 21.1.6.6, “Unsupported or Missing Features in NDB Cluster”).

**Comment by rotanid** [comment]
as i'm running the debian default configuration, this is really unexpected, shouldn't passbolt support that out of the box?
also, the storage engine is chosen bei cakephp/passbolt not by me.
besides, an index of this size sounds a bit weird, too, doesn't it?

**Comment by psy-q** [comment]
For me it's also the Debian defaults, completely unchanged. I used the install guide for Debian 8 but on Debian 9. I can reproduce it reliably on this configuration, starting from an empty VM.

**Comment by stripthis** [comment]
> as i'm running the debian default configuration, this is really unexpected, shouldn't passbolt support that out of the box?

While we test passbolt on multiple platforms there a multiple ways to configure the different components such as sessions, so we cannot test them all. For example this problems appears MariaDB 10.1 shipped with Debian but works fine with MariaDB 10.2.

>  An index of this size sounds a bit weird, too, doesn't it?
This table is provided by the cakephp framework, this is not something we created for passbolt. I tried to do a clean install of cakephp 2.x with the same config that your and ran into the same problem when installing the session table.

*Possible mid term solution*
- Suggestion/Request a fix to cakephp 2.x maintainers
- Use another session provider (such as file or memcache). See. https://book.cakephp.org/2.0/en/development/sessions.html
- Enable innodb_large_prefix option as mentionned by @cedricalfonsi 

*Long term solution*
- We are currently upgrading to cakephp v3, that will solve this issue for good

**Comment by stripthis** [comment]
@psy-q i'm unsure why you are getting the "no table found" in healthcheck. I can reproduce the SQL error but not the rest of your report.

**Comment by rotanid** [comment]
thanks for your answer!
just wanted to report that simply choosing another session provider doesnt help, but now you edited your comment adding that there is also work to be done.
should i open a ticket at cakephp or will you do it?

**Comment by rotanid** [comment]
oh boy, you can't even use the suggested workaround.
MariaDB 10.1 uses the row_format=compact by default and there is not setting to change that (the setting was introduced in 10.2)
so the SQL statements would have to have to take care of using the correct row format dynamic or compressed, because without this, the innodb_large_prefix option isn't active.

**Comment by stripthis** [comment]
@rotanid I managed to install and run passbolt on debian 9 (even with the sql error for the session table during the install) since by default passbolt is set to use the cache engine. If passbolt do not work after running the install script you might be having another error. @cedricalfonsi will follow up with you here.

**Comment by rotanid** [comment]
@stripthis thanks for the hint, i didn't really try after the error message.

**Comment by cedricalfonsi** [comment]
On Debian 9 the default charset is utf8mb4, and cake_sessions is the only table where we don't force the charset.

looking at /etc/mysql/mariadb.conf.d/50-server.cnf
```
#
# * Character sets
#
# MySQL/MariaDB default is Latin1, but in Debian we rather default to the full
# utf8 4-byte character set. See also client.cnf
#
character-set-server  = utf8mb4
collation-server      = utf8mb4_general_ci
```

Forcing the charset in utf8 solves this issue.
In app/Config/Schema/sessions.php, add the tableParameters entry as following :
```
public $cake_sessions = array(
	'id' => array('type' => 'string', 'null' => false, 'key' => 'primary'),
	'data' => array('type' => 'text', 'null' => true, 'default' => null),
	'expires' => array('type' => 'integer', 'null' => true, 'default' => null),
	'indexes' => array('PRIMARY' => array('column' => 'id', 'unique' => 1)),
	'tableParameters' => array('charset' => 'utf8', 'collate' => 'utf8_unicode_ci', 'engine' => 'InnoDB')
);
```

I created a task on our ticket system to take care of this bug (PASSBOLT-2371).

**Comment by psy-q** [comment]
@stripthis I've now tried setting up again on Debian 8.8 to make sure that the MariaDB defaults are not the issue, and I get the "No table found" error again. But the tables are there and they are populated, I can verify this directly via SQL. Running the install again didn't help.

This is the healthcheck and the result of `SELECT * FROM roles;`. Is this a completely unrelated issue, should I open a separate one?

```
---------------------------------------------------------------
     ____                  __          ____  
    / __ \____  _____ ____/ /_  ____  / / /_ 
   / /_/ / __ `/ ___/ ___/ __ \/ __ \/ / __/ 
  / ____/ /_/ (__  |__  ) /_/ / /_/ / / /    
 /_/    \__,_/____/____/_.___/\____/_/\__/   

 Open source password manager for teams
---------------------------------------------------------------
 Healthcheck shell        
---------------------------------------------------------------

 Environment

 [PASS] PHP version 5.6.30-0+deb8u1
 [PASS] PCRE compiled with unicode support
 [PASS] The app/tmp directory is writable
 [PASS] The app/webroot/img/public directory is writable

 Config files

 [PASS] The core config file is present
 [PASS] The database config file is present
 [PASS] The email config file is present
 [PASS] The application config file is present

 Core config

 [PASS] Debug mode is off.
 [PASS] Cache is working.
 [PASS] Unique value set for security.salt
 [PASS] Unique value set for security.cipherSeed
 [PASS] Full base url is set to https://passbolt
 [FAIL] App.fullBaseUrl does not validate. https://passbolt.
  [HELP] Edit App.fullBaseUrl in app/Config/core.php
  [HELP] Select a valid domain name as defined by section 2.3.1 of http://www.ietf.org/rfc/rfc1035.txt
 [FAIL] Could not reach the /healthcheck/status with the url specified in App.fullBaseUrl
  [HELP] Check that the domain name is correct in app/Config/core.php
  [HELP] Check the network settings

 SSL Certificate

 [FAIL] SSL peer certificate does not validate
 [FAIL] Hostname does not match when validating certificates.
 [WARN] Using a self-signed certificate

 Database

 [PASS] Configured to use a supported database backend
 [PASS] The application is able to connect to the database
 [PASS] Not using a prefix for database tables
 [FAIL] No table found
  [HELP] Run the install script to install the database tables
  [HELP] su -s /bin/bash -c "./app/Console/cake install" www-data
 [FAIL] No default content found
  [HELP] Run the install script to set the default content such as roles and permission types
 [PASS] The database schema up to date.

 GPG Configuration

 [PASS] PHP GPG Module is installed and loaded
 [PASS] The server gpg key is not the default one
 [PASS] The environment variable GNUPGHOME is set to /var/www/.gnupg
 [PASS] The directory containing the keyring is writable by www-data.
 [PASS] The public key file is defined in app/config.php and readable.
 [PASS] The public key file is defined in app/config.php and readable.
 [PASS] The server key fingerprint matches the one defined in app/config.php.
 [PASS] There is a valid email id defined for the server key.

 Application configuration

 [PASS] Using latest passbolt version (1.6.4)
 [PASS] Passbolt is configured to force SSL use
 [PASS] App.fullBaseUrl is set to HTTPS
 [PASS] Selenium API endpoints are disabled.
 [PASS] Search engine robots are told not to index content.
 [PASS] Registration is closed, only administrators can add users.
 [PASS] Serving the compiled version of the javascript app
 [PASS] All email notifications will be sent.

 Development Tools (optional)

 [PASS] Phpunit is installed
 [PASS] Phpunit version is 3.7.38

  6 error(s) found. Hang in there!
```

```
+--------------------------------------+-------+----------------------------+---------------------+---------------------+
| id                                   | name  | description                | created             | modified            |
+--------------------------------------+-------+----------------------------+---------------------+---------------------+
| 23d941d5-3676-3443-afdb-aaf2456f3b49 | admin | Organization administrator | 2012-07-04 13:39:25 | 2012-07-04 13:39:25 |
| 49aad81e-4f70-3380-a92e-12292597409f | guest | Non logged in user         | 2012-07-04 13:39:25 | 2012-07-04 13:39:25 |
| 857760a6-4f9d-3f1b-a292-95b630bcf03f | root  | Super Administrator        | 2012-07-04 13:39:25 | 2012-07-04 13:39:25 |
| d1acbfc1-78d8-3e25-ad8b-7ab1eb0332dc | user  | Logged in user             | 2012-07-04 13:39:25 | 2012-07-04 13:39:25 |
+--------------------------------------+-------+----------------------------+---------------------+---------------------+
```

**Comment by stripthis** [comment]
@psy-q yes open another issue. Also please check you are not using root to install passbolt, that might be your issue.

**Comment by psy-q** [comment]
No, I was running as www-data via su (copy/pasted from [these install instructions](https://medium.com/passbolt/passbolt-on-debian-8-71-from-scratch-4438dad18908)). I will try to solve it a bit longer and will create an issue if still necessary.

**Comment by cedricalfonsi** [comment]
@psy-q can you please check #164, it could be the same problem you're encountering.

**Comment by psy-q** [comment]
Thanks, that seems to be the issue. I don't know which command could have run as root, I thought I had copy/pasted exactly from the guide. No idea how this happened.

Now I'm just running into #143, I believe :)

**Comment by cedricalfonsi** [comment]
@psy-q for that one you can try to remove your browser cookies

**Comment by cedricalfonsi** [comment]
Fixed with v1.6.5.
